From: Michal Hocko <mhocko@suse.com>
Date: Wed, 14 Jun 2017 08:17:15 +0200
Subject: mm, proc: cap the stack gap for unpopulated growing vmas
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2017-1000364

Oleg has noticed that show_map_vma has been overly eager to cut the
the vma range for growing VMAs. This wasn't a big deal with 4kB stack
gap but now that the gap is much larger we can simply get a bogus VMA
range in show_map_vma.
To quote Oleg
: On ppc PAGE_SIZE == 64K, so stack_guard_gap == 16M, the application does
: mmap(..., length=4M, ... MAP_GROWSDOWN) and /proc/pid/maps happily reports
:
:       30001000000-30000400000 rw-p 00000000 00:00 0

Let's cap the reported range and show an empty range for this peculiar
case which is what we have been doing for a long time.  Note that the
range will expand as soon as the first page fault happens on this range.

Reported-by: Jan Stancek <jstancek@redhat.com>
Signed-off-by: Michal Hocko <mhocko@suse.com>
---
 fs/proc/task_mmu.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index 5ef3e14d235f..6f38f96d725f 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -303,10 +303,10 @@ show_map_vma(struct seq_file *m, struct vm_area_struct *vma, int is_pid)
 	end = vma->vm_end;
 	if (vma->vm_flags & VM_GROWSDOWN) {
 		if (stack_guard_area(vma, start))
-			start += stack_guard_gap;
+			start = min(end, start + stack_guard_gap);
 	} else if (vma->vm_flags & VM_GROWSUP) {
 		if (stack_guard_area(vma, end))
-			end -= stack_guard_gap;
+			end = max(start, end - stack_guard_gap);
 	}
 
 	seq_setwidth(m, 25 + sizeof(void *) * 6 - 1);
-- 
2.11.0

